GC-free property cursors

rickardoberg opened the pull request:
This PR primarily provides support for GC-free property cursors. Current usage through standard Kernel API will still cause minor GC (e.g. boxing and String deserialization), but has been minimized as much as possible. One major change is that Kernel API reads no longer expose Property objects, but instead returns value or null.

As transactional layer now uses cursors only the old StoreReadLayer methods for reading data have been removed.

Cypher Kernel API layer and code generator has been updated.

rickardoberg committed:
Removed methods which are now implemented by using cursors in the transactional layer

rickardoberg committed:
StorePropertyCursor reads data lazily, avoiding reading all data, which might not be needed
Updated PropertyCursor API to allow for GC-free property reading
Added ByteBufferWritableByteChannel which is handy for calling PropertyCursor.propertyData(WritableByteChannel) in a GC-free way
Added PrimitiveIntStack as a primitive int collection (used for getPropertyKeys), and implemented PrimitiveIntIterator for it

rickardoberg committed:
Kernel API reads now no longer exposes Property, but instead returns value or null. This is to minimize GC during reads

rickardoberg committed:
Updated Cypher for latest Kernel API changes

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/DataRead.java:
```diff
@@ -116,24 +112,25 @@ long nodeGetFromUniqueIndexSeek( IndexDescriptor index, Object value ) throws In
+    PrimitiveIntIterator relationshipGetPropertyKeys( long relationshipId ) throws EntityNotFoundException;
 
-    PrimitiveIntIterator relationshipGetAllPropertiesKeys( long relationshipId ) throws EntityNotFoundException;
+    PrimitiveIntIterator graphGetPropertyKeys();
```
What graphs we are talking about in `graphGetPropertyKeys`, `graphHasProperty` and `graphGetProperty`?

rickardoberg replied:
In Neo4j there are nodes, relationships and a global graph object (hardly ever used). This is for that graph object.

davidegrohmann commented on community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala:
```diff
@@ -174,12 +175,12 @@ class CompiledExecutionResult(taskCloser: TaskCloser,
-      case n: Node => (readOperations.nodeGetAllProperties(n.getId).asScala.map(_.propertyKeyId()), readOperations.nodeGetProperty _, n.getId )
-      case r: Relationship => (readOperations.relationshipGetAllProperties(r.getId).asScala.map(_.propertyKeyId()), readOperations.relationshipGetProperty _, r.getId)
+      case n: Node => (asScala(readOperations.nodeGetPropertyKeys(n.getId)), readOperations.nodeGetProperty _, n.getId )
+      case r: Relationship => (asScala(readOperations.relationshipGetPropertyKeys(r.getId)), readOperations.relationshipGetProperty _, r.getId)
```
why not `.asScala` like it was before?

rickardoberg replied:
It's a PrimitiveIntIterator which doesn't have that, so need to use JavaConversionSupport instead. Unless there's some magic that I'm not aware of that makes it work anyway.

davidegrohmann replied:
No, no magic, it is fine as it is.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java:
```diff
@@ -224,12 +224,23 @@ public PrimitiveIntIterator nodeGetLabels( long nodeId ) throws EntityNotFoundEx
-            return Property.noNodeProperty( nodeId, propertyKeyId );
+            return false;
+        }
+        return dataRead().nodeHasProperty(statement, nodeId, propertyKeyId);
```
code style

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/api/cursor/PropertyCursor.java:
```diff
@@ -75,14 +102,19 @@ public void close()
-     * @param keyId of the property to find
-     * @return true if found
-     */
     boolean seek( int keyId );
```
Removed comment because don't want doc or because it is not true anymore?

rickardoberg replied:
No, no reason really. Can't remember why I removed that, to be honest. At some point all of these methods need more Javadoc.

burqen replied:
At some point should probably be in the PR if ever.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java:
```diff
@@ -397,37 +399,69 @@ public RelationshipIterator relationshipsGetAll( KernelStatement state )
+                while ( properties.next() )
+                {
+                    DefinedProperty property = Property.property( properties.propertyKeyId(), properties.value() );
+                    indexUpdateProperty( state, nodeId, labelId, property.propertyKeyId(), null, property );
```
why do we construct a `DefinedProperty` here, it seems unnecessary (probably not related to this PR though).

rickardoberg replied:
Yeah, that code needs to be updated, but is out of scope for this PR. There's another card for it.

davidegrohmann replied:
:+1:

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java:
```diff
@@ -864,15 +977,25 @@ public PrimitiveIntIterator nodeGetPropertyKeys( KernelStatement state, long nod
             }
 
-            return Cursors.intIterator( nodeCursor.properties(), PropertyCursor.GET_KEY_INDEX_ID );
+            PrimitiveIntStack keys = new PrimitiveIntStack( 16 );
```
can we have a constant for `16`?

rickardoberg replied:
Why? Where else would we use it?

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java:
```diff
@@ -917,13 +1043,23 @@ public PrimitiveIntIterator relationshipGetPropertyKeys( KernelStatement state,
+                }
+            }
+
+            return keys.iterator();
```
Wondering if it makes sense having a `EntityCursor` interface with the `properties()` method, so we can write this code once and use it for both nodes and relationships.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/cursor/TxPropertyCursor.java:
```diff
@@ -163,7 +200,19 @@ public void close()
+        if (seekFoundIt)
+            property = Property.property( cursor.propertyKeyId(), cursor.value() );
+        seekFoundIt = false;
+    }
```
code style

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/cursor/TxPropertyCursor.java:
```diff
@@ -150,12 +152,47 @@ public boolean seek( int keyId )
+    @Override
+    public void propertyData( WritableByteChannel channel )
+    {
+        throw new UnsupportedOperationException( "NYI" );
```
Is this intended?

rickardoberg replied:
Right now it is, yes. Writing property data is quite undefined at this point, so we need to revisit this once the new clients start using it.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java:
```diff
@@ -864,15 +977,25 @@ public PrimitiveIntIterator nodeGetPropertyKeys( KernelStatement state, long nod
 
-            return Cursors.intIterator( nodeCursor.properties(), PropertyCursor.GET_KEY_INDEX_ID );
+            PrimitiveIntStack keys = new PrimitiveIntStack( 16 );
+            try (PropertyCursor properties = nodeCursor.properties())
```
Can we extract following

``` java
            PrimitiveIntStack keys = new PrimitiveIntStack( 16 );
            try (PropertyCursor properties = cursor.properties())
            {
                while (properties.next())
                {
                    keys.push( properties.propertyKeyId() );
                }
            }

            return keys.iterator();
```

to have helper method that will return cursor properties key iterator and reuse it for example in https://github.com/rickardoberg/neo4j/blob/propertycursor/community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java#L1046

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java:
```diff
@@ -902,7 +1020,15 @@ public Property nodeGetProperty( KernelStatement state, long nodeId, int propert
             }
 
-            return Cursors.iterator( storeNodeCursor.properties(), PropertyCursor.GET_PROPERTY );
+            try ( PropertyCursor cursor = storeNodeCursor.properties() )
```
Same here, we can extract helper method to do seek and get operation

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/LegacyStorePropertyCursor.java:
```diff
@@ -0,0 +1,133 @@
+    @Override
+    public void propertyData( WritableByteChannel channel )
+    {
+        throw new UnsupportedOperationException(  );
```
is it intended?

rickardoberg replied:
This class is only retained for now (not used anywhere), so that if we find any problem with StorePropertyCursor we can easily switch back to old behaviour to compare. Will be removed eventually.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyCursor.java:
```diff
@@ -29,43 +48,66 @@
+
+    private byte[] bytes = new byte[4096];
+    private ByteBuffer buffer = ByteBuffer.wrap( bytes ).order( ByteOrder.LITTLE_ENDIAN );
+    private Bits bits = Bits.bits( 32 );
```
Can we defined constants for these numbers?

rickardoberg replied:
Where else would they be used?

davidegrohmann replied:
Even if they are only used here, giving them a name will make the code more readable.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyCursor.java:
```diff
@@ -82,15 +124,521 @@ public boolean seek( int keyId )
+    private Object getRightArray()
+    {
+       // byte[] header = data.first();
+       // byte[] bArray = data.other();
```
?

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyCursor.java:
```diff
@@ -82,15 +124,521 @@ public boolean seek( int keyId )
+        {
+            buffer.order(ByteOrder.LITTLE_ENDIAN);
+        }
+    }
```
wrong code style in many places in this file

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java:
```diff
@@ -393,7 +393,14 @@ public Object getProperty( String key ) throws NotFoundException
+                Object value = statement.readOperations().nodeGetProperty( nodeId, propertyKeyId );
+
+                if (value == null)
+                    throw new PropertyNotFoundException( propertyKeyId, EntityType.NODE, nodeId );
```
code style

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/coreapi/TxStateTransactionDataSnapshot.java:
```diff
@@ -163,30 +164,46 @@ private void takeSnapshot( ReadableTxState state, StoreReadLayer storeReadLayer
+                                removedLabels.add( new LabelEntryView( nodeId, storeReadLayer.labelGetName( labels.getLabel() ) ) );
+                            }
+                        }
+                    }
```
code style

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/coreapi/TxStateTransactionDataSnapshot.java:
```diff
@@ -275,36 +292,48 @@ protected Relationship underlyingObjectToObject( Long id )
+            {
+                if (properties.seek( property ))
+                    return properties.value();
+            }
```
code style

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/coreapi/TxStateTransactionDataSnapshot.java:
```diff
@@ -275,36 +292,48 @@ protected Relationship underlyingObjectToObject( Long id )
+                if (properties.seek( property ))
+                    return properties.value();
+            }
         }
```
code style

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/PropertyStore.java:
```diff
@@ -301,7 +301,7 @@ public void ensureHeavy( PropertyBlock block )
                     {
                         stringRecords.get().setType( PropertyType.STRING.intValue() );
-                        block.addValueRecord( stringRecords.get() );
+                        block.addValueRecord( stringRecords.get().clone() );
```
why?

rickardoberg replied:
stringRecords is a cursor, so returned DynamicRecord is reused for each iteration.

davidegrohmann replied:
I see. Wondering if forcing the client to pass down a record where we can copy data in is a better design.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/PropertyStore.java:
```diff
@@ -361,9 +361,13 @@ public PropertyRecord getRecord( PropertyRecord record )
-        return storeFile.io( 0, mode );
+        PageCursor cursor =  storeFile.io( pageIdForRecord( recordId ), PF_SHARED_LOCK );
+        if (!cursor.next())
+            throw new IOException( "Record not found: "+recordId );
```
code style

MishaDemianenko replied:
Shouldn't it be `EntityNotFoundException` instead? + styling

rickardoberg replied:
On this level IOException is thrown, which will be converted to an UnderlyingStorageException. It is much more severe than an EntityNotFoundException, and indicates that something is really wrong.

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/integrationtest/PropertyIT.java:
```diff
@@ -62,10 +91,13 @@ public void shouldSetNodePropertyValue() throws Exception
-            statement.nodeSetProperty( nodeId, stringProperty( propertyKeyId, "bozo" ) );
+            statement.nodeSetProperty( nodeId,
+                    stringProperty( propertyKeyId, value ) );
+//            statement.nodeSetProperty( nodeId, byteProperty( propertyKeyId, (byte)42 ) );
```
remove commented code

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/state/IndexQueryTransactionStateTest.java:
```diff
@@ -332,7 +331,7 @@ public void before() throws Exception
         when( store.indexesGetAll() ).then( answerAsIteratorFrom( Collections.<IndexDescriptor>emptyList() ) );
         when( store.constraintsGetForLabel( labelId ) ).thenReturn( Collections.<UniquenessConstraint>emptyIterator() );
-        when( store.nodeExists( anyLong() ) ).thenReturn( true );
+//        when( store.nodeExists( anyLong() ) ).thenReturn( true );
```
remove?

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/state/LabelTransactionStateTest.java:
```diff
@@ -242,7 +237,7 @@ public void should_return_false_when_adding_existing_label() throws Exception
                 asLabelCursor( 12 ) ) );
 
-        when( store.nodeHasLabel( storeStatement, 1337, 12 ) ).thenReturn( true );
+//        when( store.nodeHasLabel( storeStatement, 1337, 12 ) ).thenReturn( true );
```
remove?

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/store/DiskLayerPropertyTest.java:
```diff
@@ -79,18 +84,38 @@ public void should_get_all_node_properties() throws Exception
                 array( 256, char.class ),
                 array( 256, float.class ),
                 array( 256, double.class ),
+*/
```
forgot to revert this back?

davidegrohmann commented:
It would be nice to have some unit tests for the new added code, e.g., `StorePropertyCursor`

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java:
```diff
@@ -971,22 +1110,27 @@ public PrimitiveIntIterator graphGetPropertyKeys( KernelStatement state )
+    public boolean graphHasProperty( KernelStatement state, int propertyKeyId )
+    {
+        return graphGetProperty( state, propertyKeyId ) != null;
+    }
```
Do we really need `graphHasProperty( KernelStatement state, int propertyKeyId )` if it doesn't provide a more efficient solution than `graphGetProperty( state, propertyKeyId )`? This will open up for something like:

```
if ( graphHasProperty( state, propertyKeyId ) )
{
    Object property = graphGetProperty( state, propertyKeyId );
}
```

Which will just be two calls for `graphGetProperty( state, propertyKeyId )`. Isn't it better to just force the usage to be the null check?

burqen replied:
Or does the API look like this to open up for more efficient implementations in the future without having to change API again?

rickardoberg replied:
Those methods are there for consistency mostly. We should revisit this in 3.0 and decide what we want to do with it.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StoreNodeRelationshipCursor.java:
```diff
@@ -76,7 +76,7 @@ public StoreNodeRelationshipCursor init( boolean isDense,
 
-        if ( isDense )
+        if ( isDense && firstRelId != -1)
         {
```
Code style and can we have a constant for -1?

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyCursor.java:
```diff
@@ -82,15 +124,521 @@ public boolean seek( int keyId )
+    {
+        if (type == null)
+            throw new IllegalStateException(  );
+
```
Code style

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyCursor.java:
```diff
@@ -82,15 +124,521 @@ public boolean seek( int keyId )
+    public Object value()
+    {
+        if (type == null)
+            throw new IllegalStateException(  );
```
Code style

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyCursor.java:
```diff
@@ -82,15 +124,521 @@ public boolean seek( int keyId )
+    public long longValue()
+    {
+        if (type == null)
+            throw new IllegalStateException(  );
```
Code style

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyCursor.java:
```diff
@@ -82,15 +124,521 @@ public boolean seek( int keyId )
+    public double doubleValue()
+    {
+        if (type == null)
+            throw new IllegalStateException(  );
```
Code style

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/LongerShortString.java:
```diff
@@ -764,15 +762,19 @@ private static void writeHeader( Bits bits, int keyId, int encoding, int stringL
-        Bits bits = Bits.bitsFromLongs( copyOf( block.getValueBlocks(),
-                block.getValueBlocks().length ) );
+        Bits bits = Bits.bitsFromLongs( block.getValueBlocks());
+        return decode(bits);
```
Code style

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/ShortArray.java:
```diff
@@ -711,6 +711,11 @@ private static void writeHeader( int keyId, ShortArray type, int arrayLength, in
     public static Object decode( PropertyBlock block )
     {
         Bits bits = Bits.bitsFromLongs(Arrays.copyOf(block.getValueBlocks(), block.getValueBlocks().length));
+        return decode(bits);
```
Code style

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/util/ByteBufferWritableByteChannel.java:
```diff
@@ -0,0 +1,98 @@
+    public int write( ByteBuffer src ) throws IOException
+    {
+        if (closed)
+            throw new ClosedChannelException();
```
Code style

MishaDemianenko replied:
Can we throw `IllegalStateException` instead? `ClosedChannelException` is a bit misleading.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/util/ByteBufferWritableByteChannel.java:
```diff
@@ -0,0 +1,98 @@
+ * Instances can be reused by calling {@link #reset()}. The internal buffer will not be reset
+ * to the initial size, but will retain whatever size it had before resetting.
+ */
+public class ByteBufferWritableByteChannel
```
Is this used somewhere now?

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/integrationtest/PropertyIT.java:
```diff
@@ -36,25 +36,54 @@
+            // THEN
+            Object value = statement.nodeGetProperty( nodeId, propertyKeyId );
+
+            System.out.println( value );
```
Should not be here

burqen commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/store/DiskLayerPropertyTest.java:
```diff
@@ -42,7 +46,7 @@ public void should_get_all_node_properties() throws Exception
         String longString =
                 "AlalalalalongAlalalalalongAlalalalalongAlalalalalongAlalalalalongAlalalalalongAlalalalalongAlalalalalong";
```
Shouldn't it be "AlalalalalongAlalalalalonglonglilonglonglongAlalalalalongAlalalalalonglonglilonglonglong"?

rickardoberg replied:
I AGREE!

burqen commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/store/DiskLayerPropertyTest.java:
```diff
@@ -79,18 +84,38 @@ public void should_get_all_node_properties() throws Exception
+                        fail();
+                    }
+                }
+            }
```
Code style

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/ConstraintEnforcingEntityOperations.java:
```diff
@@ -88,10 +88,11 @@ public boolean nodeAddLabel( KernelStatement state, long nodeId, int labelId )
+            if ( value != null )
             {
-                validateNoExistingNodeWithLabelAndProperty( state, labelId, (DefinedProperty) property, nodeId );
+                DefinedProperty property = Property.property( propertyKeyId, value );
```
I think we can avoid creation of `DefinedProperty` and modify `validateNoExistingNodeWithLabelAndProperty` to accept `propertyKeyId` and `value` separately.

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/LookupFilter.java:
```diff
@@ -110,7 +110,8 @@ public boolean test( long nodeId )
         {
-            return readOperations.nodeGetProperty( state, nodeId, propertyKeyId );
+            Object value = readOperations.nodeGetProperty( state, nodeId, propertyKeyId );
+            return value == null ? Property.noNodeProperty( nodeId, propertyKeyId ) : Property.property( propertyKeyId, value );
```
Feels like it would reasonable to remove `Property` objects creation from `LookupFilter` as part of this PR. Otherwise those objects would still be created for unique index lookups and CC here.

rickardoberg replied:
Problem is that we need valueEquals from Property. To remove these Property objects all of that would have to be worked into a separate class

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java:
```diff
@@ -176,7 +180,8 @@ public RelationshipCursor relationshipCursorGetAll( KernelStatement statement )
     {
         // TODO Filter this properly
-        return statement.getStoreStatement().acquireIteratorNodeCursor( storeLayer.nodesGetForLabel( statement, labelId ) );
+        return statement.getStoreStatement().acquireIteratorNodeCursor(
```
Is wrapping of iterator with cursor a temporary/intermediate measure?

rickardoberg replied:
yeah, I think longterm we will shift everything over to use cursors instead of iterators. No rush though.

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java:
```diff
@@ -1651,4 +1738,53 @@ public String relationshipLegacyIndexRemoveConfiguration( KernelStatement statem
+        if ( !entityCursor.next() )
+        {
+            throw new EntityNotFoundException( type, id );
+        }
```
Entity existence checks like this are very common in this class. Maybe extract method?

lutovich commented:
Removal of `Property` objects from write path (`nodeSetProperty` and `relationshipSetProperty`) and from `TxPropertyCursor` is in scope for another PR, right?

Update: never mind me

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java:
```diff
@@ -586,9 +587,13 @@ protected boolean fetchNext()
+                    if ( propertyValue != null )
                     {
-                        return next( nextValue );
+                        if ( Property.property( propertyKeyId, propertyValue ).valueEquals( value ) )
```
`propertyValue.equals( value )` ?

Update: no, I think `Objects#deepEquals` should be used because values could be arrays.

lutovich commented on community/primitive-collections/src/main/java/org/neo4j/collection/primitive/PrimitiveIntStack.java:
```diff
@@ -62,7 +62,22 @@ public void close()
+            @Override
+            public int next()
+            {
+                return array[idx++];
```
Imho check like: `if( !hasNext() ) throw new NoSuchElementException();` is required

rickardoberg replied:
Fixed!

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyCursor.java:
```diff
@@ -29,43 +48,68 @@
+    private long currentRecordId;
+    private long originalHeader;
 
     public StorePropertyCursor( PropertyStore propertyStore,
```
This class needs a unit test.

rickardoberg replied:
True. Right now it is already tested as part of the other tests (e.g. PropertyIT), but once Cypher and remoting starts using it more explicitly the unit tests need to be fleshed out

rickardoberg committed:
Updates for mandatory property constraints.

Ensure that necessary ops can be used even without a KernelStatement. This will be expanded in the future to other operations also able to work with smaller dependencies (txstate+store statement).
